Add scale-from-zero example (#1372)
networkservicemesh/deployments-k8s: examples/features/scale-from-zero/README.md

# Test automatic scale from zero

This example shows that NSEs can be created on the fly, allowing effective scaling.

## Run

Create test namespace:
```bash
NAMESPACE=($(kubectl create -f ../namespace.yaml)[0])
NAMESPACE=${NAMESPACE:10}
```

Register namespace in `spire` server:
```bash
kubectl exec -n spire spire-server-0 -- \
/opt/spire/bin/spire-server entry create \
-spiffeID spiffe://example.org/ns/${NAMESPACE}/sa/default \
-parentID spiffe://example.org/ns/spire/sa/spire-agent \
-selector k8s:ns:${NAMESPACE} \
-selector k8s:sa:default
```

Create customization file:
```bash
cat > kustomization.yaml <<EOF
---
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization

namespace: ${NAMESPACE}

bases:
- ../../../apps/nse-supplier-k8s
- ../../../apps/nsc-kernel

resources:
- supplier-role.yaml

configMapGenerator:
- name: supplier-pod-template-configmap
  files:

>   - pod-template.yaml

Reviewer: Can we also move this into `apps/nse-supplier-k8s/kustomization.yaml`?

Author: We can.
I placed the template in the example folder because I feel like it's a part of the example, and if we had a different example using the supplier, we would probably want to use another pod template.
I guess it's just a matter of preference: to have a default pod template, or to force the person, who uses supplier, always provide their own template.

Author: done

# Test automatic scale from zero

> This example shows that NSEs can be created on the fly, allowing effective scaling.

Reviewer: ```suggestion
This example shows that NSEs can be created on the fly, allowing effective scaling by the node.
```

# Test automatic scale from zero

> This example shows that NSEs can be created on the fly, allowing effective scaling by the node.

Reviewer: ```suggestion
This example shows that NSEs can be created on the fly on NSC requesting (see step 5) . That allowing effective scaling endpoints. The requested endpoint will be automatically presented on the same node with NSC that allowing the best performance for connectivity (TODO add step N).
When NSC deleting the created NSE also deleting, that saves cluster resources (see step 9 and 10).
```

## Run

> Create test namespace:

Reviewer: Let's also add steps to point them in description :)
```suggestion
1. Create test namespace:
```

```bash
NAMESPACE=($(kubectl create -f ../namespace.yaml)[0])
NAMESPACE=${NAMESPACE:10}
```

> Register namespace in `spire` server:

Reviewer: ```suggestion
2. Register namespace in `spire` server:
```

```bash
kubectl exec -n spire spire-server-0 -- \
/opt/spire/bin/spire-server entry create \
-spiffeID spiffe://example.org/ns/${NAMESPACE}/sa/default \
-parentID spiffe://example.org/ns/spire/sa/spire-agent \
-selector k8s:ns:${NAMESPACE} \
-selector k8s:sa:default
```

> Create customization file:

Reviewer: ```suggestion
3. Create kustomization file:
```

```bash
cat > kustomization.yaml <<EOF
---
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization

namespace: ${NAMESPACE}

bases:
- ../../../apps/nse-supplier-k8s
- ../../../apps/nsc-kernel

patchesStrategicMerge:
- patch-nsc.yaml
- patch-supplier.yaml
EOF
```

> Register network service:

Reviewer: ```suggestion
4. Register network service:
```

```bash
kubectl apply -f scale-ns.yaml
```

> Deploy NSC and supplier:

Reviewer: ```suggestion
5. Deploy NSC and supplier:
```

```bash
kubectl apply -k .
```

Wait for applications ready:
```bash

> kubectl wait --for=condition=ready --timeout=1m pod -l app=nse-icmp-responder -n ${NAMESPACE}

Reviewer: Why do we need to wait?

Author: On a fresh cluster start of the pods can time quite long time. If we don't wait, subsequent commands will fail.

Reviewer: we have a prefetch for this. This should not affect the test.

Author: All of the other tests have wait till pods are ready. Are you sure it's safe to remove the wait?
Also, if someone replicates this example manually, they will have to wait manually until all the images are downloaded.

This wait doesn't increase duration of the test. If we already have everything in the cache, and pods started immediately, `kubectl wait --for=condition=ready` will return without any delay.

> Wait for applications ready:
> ```bash
> kubectl wait --for=condition=ready --timeout=1m pod -l app=nse-icmp-responder -n ${NAMESPACE}
> ```

Reviewer: ```suggestion
```

> Find NSC and NSE pods by labels:

Reviewer: ```suggestion
6. Find NSC and NSE pods by labels:
```

```bash
NSC=$(kubectl get pods -l app=nsc-kernel -n ${NAMESPACE} --template '{{range .items}}{{.metadata.name}}{{"\n"}}{{end}}')
```
```bash
NSE=$(kubectl get pods -l app=nse-icmp-responder -n ${NAMESPACE} --template '{{range .items}}{{.metadata.name}}{{"\n"}}{{end}}')
```

> Check connectivity:

Reviewer: ```suggestion
7. Check connectivity from the client:
```

```bash
kubectl exec ${NSC} -n ${NAMESPACE} -- ping -c 4 169.254.0.0
```

> Check connectivity:

Reviewer: ```suggestion
8. Check connectivity from the endpoint:
```

```bash
kubectl exec ${NSE} -n ${NAMESPACE} -- ping -c 4 169.254.0.1
```

> Remove the client pod:

Reviewer: ```suggestion
9. Remove the client pod:
```

```bash
kubectl scale -n ${NAMESPACE} deployment nsc-kernel --replicas=0
```

> Wait for the NSE pod to be deleted:

Reviewer: ```suggestion
10. Wait for the NSE pod to be deleted:
```
